Enabled add, sum and mul for bool tensor

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21033 Enable all and any for bool tensors
* **#21032 Enabled add, sum, mul and neg for bool tensor**
* #21031 Enabled min, max, minall, maxall, cmin, cmax, cmaxValue, cminValue for bool tensors
* #21030 Added possibility to index scalars by bool masks

This PR is a part of a stack which will change result tensor type of comparison ops from uint8 to bool. As this change is rather big and a lot of prep work is needed, im breaking it into a stack.

Changes in this PR:
- Enable add, sum, mul and neg for bool tensors.

Differential Revision: [D15530496](https://our.internmc.facebook.com/intern/diff/D15530496)

izdeby committed:
Enabled add, sum, mul and neg for bool tensor

colesbury commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -29,12 +29,22 @@ void sub_kernel(TensorIterator& iter, Scalar alpha_scalar) {
 }
 
+template <typename T>
+T multiply(T a, T b) {
```
Try to avoid this. `*` should work on `bool`. You might need to add an overload to Vec256.

izdeby replied:
It would work without this but the build on CircleCi is failing due to an error about bad usage of "*" with two bools.

colesbury replied:
Then suppress the warning. I'm not a fan of unnecessarily verbose code driven by the compiler's preferences.

izdeby replied:
@cpuhrsch, made a good argument against adding an overload to Vec256.
Also, i dont feel comfortable with fully blocking warning for cases like this. If there was a way to block it for this method, i would be fine but as far as i can tell - we block gcc warnings for the whole project and i dont think its a good way to go. 
Please, let me know if you have strong preferences here

colesbury replied:
```
#pragma GCC diagnostic ignored "-Wint-in-bool-context"
```

We should stick to the same operators because it keeps the code simple and **reduces** the chances of incorrect behavior. For example, this PR previously had an incorrect definition of `-` because it didn't use `operator-` -- it was the opposite of how `-` operates on a bool in both C/C++ and Python.

colesbury commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -92,12 +92,19 @@ static void reciprocal_kernel(TensorIterator& iter) {
 
 static void neg_kernel(TensorIterator& iter) {
-  AT_DISPATCH_ALL_TYPES(iter.dtype(), "neg_cpu", [&]() {
+  if (iter.dtype() == ScalarType::Bool) {
```
This doesn't look right.  We shouldn't implement unary `-` on bool. We should implement bitwise not (unary `~`) instead.

It's also worth noting that this is the opposite of C++ behavior. For example:

```
bool x = ...
bool y = -x;
```

x and y have the same value (because 0=>0=>false and 1=>-1=>true)

izdeby replied:
removing this change for now. will implement this as a separate PR

colesbury commented on aten/src/THC/THCNumerics.cuh:
```diff
@@ -66,12 +66,26 @@ struct THCNumerics<uint8_t> {
-  static inline __host__ __device__ bool ge(uint8_t a, uint8_t b) { return a >= b; }
-  static inline __host__ __device__ bool eq(uint8_t a, uint8_t b) { return a == b; }
-  static inline __host__ __device__ bool ne(uint8_t a, uint8_t b) { return a != b; }
+  static inline __host__ __device__ bool min() { return at::numeric_limits<bool>::lowest(); }
```
Where is this used? We shouldn't be using THCNumerics in these functions any more and should try to avoid adding anything to it.

izdeby replied:
Please see the [build error ](https://circleci.com/gh/pytorch/pytorch/1817362?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) for a previous PR in the stack. I moved numerics changes to the previous PR.

colesbury reviewed (commented):
This needs a test for sum(). Where are the modifications for CUDA?

izdeby committed:
Update on "Enabled add, sum, mul and neg for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "[WIP] Enabled add, sum, mul and neg for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "[WIP] Enabled add, sum, mul and neg for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

ezyang commented:
Don't forget to fix lint

VitalyFedyunin commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -307,6 +307,14 @@ template <class T> Vec256<T> inline operator*(const Vec256<T> &a, const Vec256<T
+template <> Vec256<bool> inline operator*(const Vec256<bool> &a, const Vec256<bool> &b) {
+  Vec256<bool> c = Vec256<bool>();
+  for (int i = 0; i != Vec256<bool>::size(); i++) {
+    c[i] = a[i] && b[i];
```
Not sure if it is going to work in your case, but please consider checking https://scc.ustc.edu.cn/zlsc/sugon/intel/compiler_c/main_cls/index.htm#intref_cls/common/intref_avx_fmadd_pd.htm

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "[WIP] Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "[WIP] Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby commented:
@pytorchbot retest this please

izdeby committed:
Update on "[WIP] Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby commented:
@pytorchbot retest this please

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

izdeby committed:
Update on "Enabled add, sum and mul for bool tensor"

Enabled add, sum, mul and neg for bool tensor

gh-metadata: pytorch pytorch 21032 gh/izdeby/3/head

gchanan commented:
this added support for subtraction with bools (by accident?) but it's not tested anywhere.
